### PR TITLE
Puppet support in Federated Replica Set

### PR DESCRIPTION
**Pull request by mwielgus**
See: https://github.com/kubernetes/community/pull/325 for details.

cc: @nikhiljindal @csbell @madhusudancs

### CODE REVIEW

**Comment by mwielgus** [comment]
@k8s-bot gci gce e2e test this

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go [hunk 1]
```diff
@@ -567,7 +570,7 @@ func (frsc *ReplicaSetController) reconcileReplicaSet(key string) (reconciliatio
 		lrs.Spec.Replicas = &specReplicas
 
 		if !exists {
-			if replicas > 0 {
+			if replicas > 0 && !isPuppet(frs) {
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: I see this isPuppet sprinkled at many places in this function.
How about refactoring this function to create 2 smaller functions for:
* Updating cluster replicas spec as per federated replicaset; and
* Updating federated replicaset status as per cluster replicas.

This function (reconcileReplicaset) can call both these smaller functions.
We can skip calling the first one and call the second one for puppet replicasets. The 2 smaller functions dont need to know about puppet replicasets.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go [hunk 2]
```diff
@@ -580,24 +583,43 @@ func (frsc *ReplicaSetController) reconcileReplicaSet(key string) (reconciliatio
...
-	if fedStatus.Replicas != frs.Status.Replicas || fedStatus.FullyLabeledReplicas != frs.Status.FullyLabeledReplicas ||
-		fedStatus.ReadyReplicas != frs.Status.ReadyReplicas || fedStatus.AvailableReplicas != frs.Status.AvailableReplicas {
+	if fedStatus.Replicas != frs.Status.Replicas ||
+		fedStatus.FullyLabeledReplicas != frs.Status.FullyLabeledReplicas ||
+		fedStatus.ReadyReplicas != frs.Status.ReadyReplicas ||
+		fedStatus.AvailableReplicas != frs.Status.AvailableReplicas ||
+		(isPuppet(frs) && *frs.Spec.Replicas != localSpecReplicaSum) {
+
+		if isPuppet(frs) {
+			frs.Spec.Replicas = &localSpecReplicaSum
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: Why only update spec.Replicas.
Dont we need to update status as well?

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go [hunk 3]
```diff
@@ -180,6 +175,81 @@ func TestReplicaSetController(t *testing.T) {
 	assert.Equal(t, rs.Status.AvailableReplicas, rs1.Status.AvailableReplicas+rs2.Status.AvailableReplicas)
 }
 
+func TestReplicaSetControllerPuppets(t *testing.T) {
+
+	replicaSetReviewDelay = 10 * time.Millisecond
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: Is this just a copy of lines 73 above?

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go [hunk 4]
```diff
@@ -180,6 +175,81 @@ func TestReplicaSetController(t *testing.T) {
...
+	clusterAvailableDelay = 20 * time.Millisecond
+	clusterUnavailableDelay = 60 * time.Millisecond
+	allReplicaSetReviewDelay = 120 * time.Millisecond
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: I dont see this being used anywhere. What am i missing?

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go [hunk 5]
```diff
@@ -180,6 +175,81 @@ func TestReplicaSetController(t *testing.T) {
...
+
+	fedclientset := fedclientfake.NewSimpleClientset()
+	fedrswatch := watch.NewFake()
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: fedrswatch is really confusing. How about naming it fedRsWatch?

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go [hunk 6]
```diff
@@ -180,6 +175,81 @@ func TestReplicaSetController(t *testing.T) {
...
+	fedclientset.PrependWatchReactor("replicasets", core.DefaultWatchReactor(fedrswatch, nil))
+
+	fedclientset.Federation().Clusters().Create(testutil.NewCluster("k8s-1", apiv1.ConditionTrue))
+	fedclientset.Federation().Clusters().Create(testutil.NewCluster("k8s-2", apiv1.ConditionTrue))
+
+	kube1clientset := kubeclientfake.NewSimpleClientset()
+	kube1rswatch := watch.NewFake()
+	kube1clientset.PrependWatchReactor("replicasets", core.DefaultWatchReactor(kube1rswatch, nil))
+	kube1Podwatch := watch.NewFake()
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: name it kube1PodWatch?

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go (hunk 5 above)
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: similar comment for other variables here

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go [hunk 7]
```diff
@@ -180,6 +175,81 @@ func TestReplicaSetController(t *testing.T) {
...
+func TestReplicaSetControllerPuppets(t *testing.T) {
+
+	replicaSetReviewDelay = 10 * time.Millisecond
+	clusterAvailableDelay = 20 * time.Millisecond
+	clusterUnavailableDelay = 60 * time.Millisecond
+	allReplicaSetReviewDelay = 120 * time.Millisecond
+
+	fedclientset := fedclientfake.NewSimpleClientset()
+	fedrswatch := watch.NewFake()
+	fedclientset.PrependWatchReactor("replicasets", core.DefaultWatchReactor(fedrswatch, nil))
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: Why is this watch reactor required?

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go [hunk 8]
```diff
@@ -180,6 +175,81 @@ func TestReplicaSetController(t *testing.T) {
...
+
+	rs2 := newReplicaSetWithReplicas("rs", 2)
+	rs2.Status.Replicas = *rs2.Spec.Replicas
+	rs2.Status.FullyLabeledReplicas = *rs2.Spec.Replicas
+	rs2.Status.ReadyReplicas = *rs2.Spec.Replicas
+	rs2.Status.AvailableReplicas = *rs2.Spec.Replicas
+	rs2, _ = kube2clientset.Extensions().ReplicaSets(metav1.NamespaceDefault).Create(rs2)
+	kube2rswatch.Add(rs2)
+
+	time.Sleep(1 * time.Second)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: Sleep seems flaky. Replace by a polling to see Get does not return an error? Or add a watch?

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go [hunk 9]
```diff
@@ -180,6 +175,81 @@ func TestReplicaSetController(t *testing.T) {
...
+	rs, _ = fedclientset.Extensions().ReplicaSets(metav1.NamespaceDefault).Get(rs.Name, metav1.GetOptions{})
+	assert.Equal(t, *rs1.Spec.Replicas+*rs2.Spec.Replicas, rs.Status.Replicas)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: Shouldnt the right most argument be rs.Spec.Replicas?

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go [hunk 10]
```diff
@@ -180,6 +175,81 @@ func TestReplicaSetController(t *testing.T) {
 	assert.Equal(t, rs.Status.AvailableReplicas, rs1.Status.AvailableReplicas+rs2.Status.AvailableReplicas)
 }
 
+func TestReplicaSetControllerPuppets(t *testing.T) {
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: Will be great if you can add a comment here summarizing what this test does.
Creates a puppet federated RS and 2 RSs in underlying clusters and then verifies that the federated RS spec and status are updated as per the spec and status from RSs in underlying clusters.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go [hunk 11]
```diff
@@ -55,6 +56,7 @@ const (
 	FedReplicaSetPreferencesAnnotation = "federation.kubernetes.io/replica-set-preferences"
 	allClustersKey                     = "THE_ALL_CLUSTER_KEY"
 	UserAgentName                      = "Federation-replicaset-Controller"
+	FedReplicaSetPupptetAnnotation     = "federation.kubernetes.io/is-puppet"
```
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: Done.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go (hunk 2 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: Status is updated always, spec only for replicas.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go (hunk 3 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: It is needed to increase controller responsiveness. These are global variables. Discussed offline.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go (hunk 4 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: See above.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go (hunk 5 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: I'm for rewriting this test from scratch/significantly refactor. I would be easier if these two tests use the same names.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go (hunk 9 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: Yes. Done.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go (hunk 8 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: Done.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go (hunk 6 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: Done. Renamed also other variables in the whole file.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go (hunk 7 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: DefaultWatchReactor copies all updates to cache.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go (hunk 5 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go]: Renamed all variables.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go (hunk 1 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: I did some refactoring, however i would rather keep all the logic in the single place, no separate functions for puppet reconcilation etc.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go [hunk 12]
```diff
@@ -580,24 +586,33 @@ func (frsc *ReplicaSetController) reconcileReplicaSet(key string) (reconciliatio
...
-			fedStatus.FullyLabeledReplicas += currentLrs.Status.FullyLabeledReplicas
-			fedStatus.ReadyReplicas += currentLrs.Status.ReadyReplicas
-			fedStatus.AvailableReplicas += currentLrs.Status.AvailableReplicas
 		}
 	}
-	if fedStatus.Replicas != frs.Status.Replicas || fedStatus.FullyLabeledReplicas != frs.Status.FullyLabeledReplicas ||
-		fedStatus.ReadyReplicas != frs.Status.ReadyReplicas || fedStatus.AvailableReplicas != frs.Status.AvailableReplicas {
+	if hasStatusChanged(&frs.Status, &fedStatus) || (isPuppet(frs) && *frs.Spec.Replicas != localSpecReplicaSum) {
+		if isPuppet(frs) {
+			frs.Spec.Replicas = &localSpecReplicaSum
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: We will need another call to Update the spec. UpdateStatus below will only update status

**Comment by nikhiljindal** [comment]
mostly lgtm. Need to fix the update and update status issue as discussed

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go (hunk 12 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: Done.

**Comment by marun** [comment]
Review status: 0 of 2 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go, line 83 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40836#-KcPau1gWQ8j8XvkfK2B:-KcPau1gWQ8j8XvkfK2C:b6ksspq) ([raw file](https://github.com/kubernetes/kubernetes/blob/0bb31560c21d891966685c6b829af628a4486ed0/federation/pkg/federation-controller/replicaset/replicasetcontroller_test.go#L83)):*
> ```Go
> 	allReplicaSetReviewDelay = 120 * time.Millisecond
> 
> 	fedclientset := fedclientfake.NewSimpleClientset()
> ```

Consider breaking non-functional formatting changes such as this into one or more separate commits.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40836)*

**Comment by marun** [comment]
Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40836)*

**Pull request closed without merging**